Unselect empty groups during shorthand creation

jan-cerny opened the pull request:
Remove Python script `unselect_empty_xccdf_groups.py`. This script
modified every profile in the given XCCDF file by unselecting XCCDF
Groups that don't contain any rules selected by the profile. Instead of
using this script this job will be done during shorthand creation
by the `build_shorthand.py` tool.

This helps to reduce amount of steps during build process and prevents
unnecessary loading and saving. Moreover, the script
`unselect_empty_xccdf_groups.py` worked only with XCCDF 1.1, this change
will help us to phase out XCCDF 1.1 from the build system.

jan-cerny commented:
I have rebased this on the top of the latest upstream master branch, because the required tests in GitHub CI have changed.

jan-cerny commented:
I have just realized that during then XML serialization of profiles it isn't possible to distinguish rule selections from profile selections. At the current situation we don't care because we generate XCCDF 1.1 where we use the short ID and when converting  XCCDF 1.1 to 1.2 later the correct long ID is added later by the oscap's XSLT. But, in future, when we would like to generate XCCDF 1.2 directly, we will need to add either xccdf_org.ssgproject_content_rule_  or xccdf_org.ssgproject_content_group prefix during serialization of profile selections to XML depending whether the selection is a rule selection or a group selection.

We already distinguish variables because they have a `=` in the selections. But, group IDs are not special in any way, they are just arbitrary [a-z_] strings.

The only idea that comes to my mind to solve this issue is to have a separate `groups:` key in the resolved profile YAMLs. Then adapt the respective classes (Profile and SelectionHandler).

Do you have any other ideas or suggestions? What is your opinion?

yuumasato reviewed (commented):
> But, in future, when we would like to generate XCCDF 1.2 directly, we will need to add either xccdf_org.ssgproject_content_rule_ or xccdf_org.ssgproject_content_group prefix during serialization of profile selections to XML depending whether the selection is a rule selection or a group selection.

@jan-cerny So you think that when we get closer to building XCCDF 1.2 directly and the Rule and Group loader add the XCCDF rule or group prefix and we won't need the `unselected_groups` key in the compiled profiles?

I'm not a fan of having all the unselected groups in the compiled profile and stability tests, but I'm fine if this will be a temporary solution.

jan-cerny commented:
> So you think that when we get closer to building XCCDF 1.2 directly and the Rule and Group loader add the XCCDF rule or group prefix and we won't need the unselected_groups key in the compiled profiles?

I'm sorry it's confusing. I meant the opposite. I'm adding the `unselected_groups` key because I have in mind the change to build XCCDF 1.2 directly in future.

Actually, for current system where we build XCCDF 1.1, it's enough to add the IDs of groups to the `selections` key, ie. mix the group IDs in between rule IDs. The generated XCCDF 1.1 file contains the IDs in the form without the prefixes and when the XSLT creates XCCDF 1.2 file from the XCCDF 1.1 file the XSLT correctly assigns prefixes to the selections, it's able to find out which of the selections are rules and which of them are groups.

Now when we will want to generate XCCDF 1.2 directly, we will have to generate XML Profile element with child select elements with the prefixes. However, the Profile class in build_yaml.py works with the short IDs and doesn't have information which of the selections are groups and which of them are rules. So my proposal separates groups into separate list, and therefore it could easily assign the correct prefix to the group IDs.

> I'm not a fan of having all the unselected groups in the compiled profile and stability tests, but I'm fine if this will be a temporary solution.

If we do it the way proposed in the PR it won't be temporary solution, it will be permanent solution.

It's done in this way because it fits more into the current logic: during resolution of profile, all selection, unselections, slection inheritance is resolved and written out to the compiled profile and in the second phase it's just loaded and converted to an XML Profile element.

Also with this change the compiled profile contents are more similar to the final contents of the XML Profile element, the group selections don't appear magically.


I will try to brainstorm alternative solutions:
- make the Profile class more aware of the context ie. to have a method in the Benchmark class that would get list of groups and list of rules and this method would be used during XML serialization
- rename all groups to start their name with group_ and the build system would easily recognize a group based on the name
- work with full IDs instead of short IDs in the whole build_shorthand.py script
- don't integrate the logic that unselects empty groups into any script, keep it as a separate build step and only adjust it to be able to consume XCCDF 1.2

Thoughts?

jan-cerny commented:
I have add a comment

yuumasato commented on build-scripts/compile_all.py:
```diff
@@ -105,6 +105,7 @@ def load_resolve_and_validate_profiles(env_yaml, profile_files, loader, controls
         p.validate_variables(loader.all_values.values())
         p.validate_rules(loader.all_rules.values(), loader.all_groups.values())
         p.validate_refine_rules(loader.all_rules.values())
+        p.unselect_empty_groups(loader.loaded_group)
```
How about calling `unselect_empty_groups()` after we `load_compiled_content()` in `build_shorthand.py`?
Maybe in `load_benchmark()`?

yuumasato reviewed (commented):
> It's done in this way because it fits more into the current logic: during resolution of profile, all selection, unselections, selection inheritance is resolved and written out to the compiled profile and in the second phase it's just loaded and converted to an XML Profile element.

I understand the will to be more transparent, but I don't see much value in the `unselected_groups` to make them so visible. (I think it would make sense to have a test though, to make sure we are only unselecting empty groups)
I  think the compiled profiles should reflect and make explicit what our inputs in rules, profiles, groups and templates will translate into. The unselected groups is an indirect effect of the rules we select or not, we don't have direct control over them.

> 
> Also with this change the compiled profile contents are more similar to the final contents of the XML Profile element, the group selections don't appear magically.
> 
> I will try to brainstorm alternative solutions:
> 
>     * make the Profile class more aware of the context ie. to have a method in the Benchmark class that would get list of groups and list of rules and this method would be used during XML serialization

Instead of making Profile class more aware, how about shifting the moment when the empty groups are unselected? Like doing it in `build_shorthand.py` instead of in `compile_all.py`?
Somewhere between `load_compiled_content()` and `load_benchmark()`...

> 
>     * rename all groups to start their name with group_ and the build system would easily recognize a group based on the name
> 
>     * work with full IDs instead of short IDs in the whole build_shorthand.py script

In build_shorthand.py we can infer the type of the content by its class.
But for the profiles the IDs are just strings, so I guess that at some point we will need to have a way to discover if an ID is a Rule or a Group. I guess the IDs of Rules and Groups should not collide, so there could be a dictionary to map ID to type of content.

>     * don't integrate the logic that unselects empty groups into any script, keep it as a separate build step and only adjust it to be able to consume XCCDF 1.2

Nah.

> 
> 
> Thoughts?

jan-cerny commented:
great!

yuumasato reviewed (commented):
Looks great to me.

But @jan-cerny, would you like to simplify `_find_empty_groups()`?
Codeclimate says it is a bit more complex than what we set the limits to.

yuumasato reviewed (approved):
I am, personally, okay with the current state of `_find_empty_groups()`

jan-cerny commented:
I see that you already discovered the latest changes :smile: . Thanks for your idea, I have reworked the PR to perform the unselection during creation of shorthand.xml rather than during profile resolution. I have discovered that the Benchmark class actually has access to both profiles and groups so I have moved the entry point there.

>But @jan-cerny, would you like to simplify _find_empty_groups()?
Codeclimate says it is a bit more complex than what we set the limits to.

I was looking into it last week and I couldn't figure anything. It's probably acceptable to keep it as is because it's quite isolated algorithm.

jan-cerny commented:
@ggbecker Nice catch! I think that it is a conflict between #9332 and #9345, I will go fix it in #9436.
